File: web/sites/domain/vas/vasworld.php
<?php
	class VASWorld
	{
		public	$id,
				//$partner_id,
				$agreement_id,
				$agreement,
				$name,
				$content,
				$status,
				$remarks,
				$date_created,
				$date_updated;

		public static $STATUS_DRAFT = 0;
		public static $STATUS_PENDING = 1;
		public static $STATUS_REJECTED = 2;
		public static $STATUS_CANCELLED = 3;
		public static $STATUS_WITHDREW = 4;
		public static $STATUS_APPROVED = 5;
		public static $STATUS_PUBLISHED = 6;

		public function __construct($data)
		{
			$this->id = get_value_from_array('ID', $data);
			//$this->partner_id = get_value_from_array('PartnerID', $data);
			$this->agreement_id = get_value_from_array('AgreementID', $data);
			$this->name = get_value_from_array('Name', $data);
			$this->content = get_value_from_array('Content', $data);
			$this->status = get_value_from_array('Status', $data);
			$this->remarks = get_value_from_array('Remarks', $data);
			$this->date_created = get_value_from_array('DateCreated', $data);
			$this->date_updated = get_value_from_array('DateUpdated', $data);
			$this->agreement = null;
		}

		public function status()
		{
			switch($this->status)
			{
				case self::$STATUS_DRAFT: $status = 'Draft'; break;
				case self::$STATUS_WITHDREW: $status = 'Withdrawn'; break;
				case self::$STATUS_CANCELLED: $status = 'Cancelled'; break;
				case self::$STATUS_PENDING: $status = 'Pending Approval'; break;
				case self::$STATUS_REJECTED: $status = 'Rejected'; break;
				case self::$STATUS_APPROVED: $status = 'Approved'; break;
				case self::$STATUS_PUBLISHED: $status = 'Published'; break;
			}

			return $status;
		}

		public function next_action()
		{
			switch($this->status)
			{
				case self::$STATUS_DRAFT: $action = 'Submit'; break;
				case self::$STATUS_WITHDREW: $action = null; break;
				case self::$STATUS_CANCELLED: $action = null; break;
				case self::$STATUS_PENDING: $action = 'Withdraw'; break;
				case self::$STATUS_REJECTED: $action = null; break;
				case self::$STATUS_APPROVED: $action = 'Publish'; break;
				case self::$STATUS_PUBLISHED: $action = null; break;
			}

			return $action;
		}

		public function date_submitted()
		{
			return date("d M Y g:i (A)", strtotime($this->date_created));
		}

		public function date_updated()
		{
			return date("d M Y g:i (A)", strtotime($this->date_updated));
		}
	}
?>

File: mis-ror/ror/app/models/vasworld.rb
class Vasworld

    # Attributes
    @id
    @partner
    @partnerId
    @name
    @content
    @status
    @remarks
    @dateCreated
    @dateUpdated
    
    attr_accessor :id, :partner, :name, :content, :status, :remarks, :dateCreated, :dateUpdated
    
    # Constant
    STATUS_DRAFT = 0
    STATUS_PENDING = 1
    STATUS_REJECTED = 2
    STATUS_CANCELLED = 3
    STATUS_WITHDREW = 4
    STATUS_APPROVED = 5
    STATUS_PUBLISHED = 6
    
    
    def initialize(args={})
        @id = args[:id] || nil
        @partnerId = args[:partnerId] || nil

        @partner = args[:partner] || nil
        @name = args[:name] || nil
        @content = args[:content] || nil
        @status = args[:status] || nil
        @remarks = args[:remarks] || nil
        
        @dateCreated = DateTime.strptime(args[:dateCreated].to_s, '%Y-%m-%d %H:%M:%S').to_formatted_s(:long_ordinal) || nil 
        @dateUpdated = DateTime.strptime(args[:dateUpdated].to_s, '%Y-%m-%d %H:%M:%S').to_formatted_s(:long_ordinal) || nil 
    end
    
    
    def show_status
        #
        # Show status message
        #
        
        case @status
            when STATUS_DRAFT
                status = 'Draft'    
            when STATUS_WITHDREW
                status = 'Withdrew'
            when STATUS_CANCELLED
                status = 'Cancelled'
            when STATUS_PENDING
                status = 'Pending Approval'
            when STATUS_REJECTED
                status = 'Rejected'
            when STATUS_APPROVED
                status = 'Approved'
            when STATUS_PUBLISHED
                status = 'Published'
        end 
        
        return status
    end
    
    def self.create(args={})
        #
        # Create a VASWorld
        #
        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "INSERT INTO partnervasworld 
                             (Name, Content, Status, DateCreated, DateUpdated, Remarks, PartnerAgreementID) 
                             VALUES (?, ?, ?, NOW(), NOW(), ?, ?)"
        stmt.execute args[:name], args[:content], args[:status], args[:remarks], args[:agreementId]
        
        @vasworldId = stmt.insert_id    
            
        stmt.close
        
        return @vasworldId
    end
    
    def self.list(status)
        #
        # Select a list of VASWorld
        #

        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "SELECT pvw.Id, pvw.Name, pvw.Content, pvw.Status, pvw.Remarks, 
                                  pvw.DateCreated, pvw.DateUpdated, 
                                  p.Id as PartnerID, p.Name as PartnerName, p.DateCreated as PartnerDateCreated 
                             FROM partnervasworld pvw, partneragreement pa, partner p 
                            WHERE p.Id = pa.PartnerId 
                              AND pa.Id = pvw.PartnerAgreementId
                              AND pvw.Status = ? 
                         ORDER BY pvw.DateUpdated DESC"
        stmt.execute status
        
        vasworlds = []
        
        while row = stmt.fetch do
        
            partner = Partner.new(
                    :id => row[7],
                    :name => row[8],
                    :dateCreated => row[9]
                )
        
            vasworld = Vasworld.new(
                    :id => row[0],
                    :partner => partner,
                    :name => row[1],
                    :content => row[2],
                    :status => row[3],
                    :remarks => row[4],
                    :dateCreated => row[5],
                    :dateUpdated => row[6]
                )
            
            vasworlds << vasworld
        end
        
        stmt.free_result
        stmt.close
        
        return vasworlds
    end
    
    def self.past(limit=10)
        #
        # List past approved/rejected VASWorld
        #
        
        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "SELECT pvw.Id, pvw.Name, pvw.Content, pvw.Status, pvw.Remarks, 
                                  pvw.DateCreated, pvw.DateUpdated, 
                                  p.Id as PartnerID, p.Name as PartnerName, p.DateCreated as PartnerDateCreated 
                             FROM partnervasworld pvw, partneragreement pa, partner p 
                            WHERE pa.Id = pvw.PartnerAgreementId 
                              AND (pvw.Status = ? OR pvw.Status = ?)
                              AND pa.PartnerId = p.Id 
                         ORDER BY pvw.DateUpdated DESC"
        stmt.execute STATUS_APPROVED, STATUS_REJECTED
        
        vasworlds = []
        
        while row = stmt.fetch do
        
            partner = Partner.new(
                    :id => row[7],
                    :name => row[8],
                    :dateCreated => row[9]
                )
        
            vasworld = Vasworld.new(
                    :id => row[0],
                    :partner => partner,
                    :name => row[1],
                    :content => row[2],
                    :status => row[3],
                    :remarks => row[4],
                    :dateCreated => row[5],
                    :dateUpdated => row[6]
                )
            
            vasworlds << vasworld
        end
        
        stmt.free_result
        stmt.close
    
        return vasworlds
    end
    
    def self.find(id)
        #
        # Get a VASWorld
        #
        
        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "SELECT pvw.Id, pa.PartnerId, pvw.Name, pvw.Content, pvw.Status, pvw.Remarks, 
                                  pvw.DateCreated, pvw.DateUpdated 
                             FROM partnervasworld pvw, partneragreement pa 
                            WHERE pvw.Id = ? AND pa.Id = pvw.PartnerAgreementId"
        stmt.execute id
        
        while row = stmt.fetch do
            vasworld = Vasworld.new(
                    :id => row[0],
                    :partner_id => row[1],
                    :name => row[2],
                    :content => row[3],
                    :status => row[4],
                    :remarks => row[5],
                    :dateCreated => row[6],
                    :dateUpdated => row[7]
                )
        end     
        
        stmt.free_result
        stmt.close
        
        return vasworld
    end
    
    def self.approve(id)
        #
        # Approve pending VASWorld
        #
        
        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "UPDATE partnervasworld pvw 
                              SET status = ?
                            WHERE pvw.Id = ?"
        
        stmt.execute STATUS_APPROVED, id
        stmt.close
    end
    
    def self.reject(id)
        #
        # Reject pending VASWorld
        #
        
        db = Dbconn.get_instance('master')
        
        stmt = db.prepare "UPDATE partnervasworld pvw 
                              SET status = ?
                            WHERE pvw.Id = ?"
        
        stmt.execute STATUS_REJECTED, id
        stmt.close
    end
end

